File: typescript/dynamiccolor/material_dynamic_colors.ts
/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {ColorSpecDelegateImpl2025} from './color_spec_2025.js';
import {DynamicColor} from './dynamic_color.js';
import type {DynamicScheme} from './dynamic_scheme';

/**
 * DynamicColors for the colors in the Material Design system.
 */
// Material Color Utilities namespaces the various utilities it provides.
// tslint:disable-next-line:class-as-namespace
export class MaterialDynamicColors {
  static contentAccentToneDelta = 15.0;

  private static readonly colorSpec = new ColorSpecDelegateImpl2025();

  highestSurface(s: DynamicScheme): DynamicColor {
    return MaterialDynamicColors.colorSpec.highestSurface(s);
  }

  ////////////////////////////////////////////////////////////////
  // Main Palettes                                              //
  ////////////////////////////////////////////////////////////////

  primaryPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.primaryPaletteKeyColor();
  }

  secondaryPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.secondaryPaletteKeyColor();
  }

  tertiaryPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.tertiaryPaletteKeyColor();
  }

  neutralPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.neutralPaletteKeyColor();
  }

  neutralVariantPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.neutralVariantPaletteKeyColor();
  }

  errorPaletteKeyColor(): DynamicColor {
    return MaterialDynamicColors.colorSpec.errorPaletteKeyColor();
  }

  ////////////////////////////////////////////////////////////////
  // Surfaces [S]                                               //
  ////////////////////////////////////////////////////////////////

  background(): DynamicColor {
    return MaterialDynamicColors.colorSpec.background();
  }

  onBackground(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onBackground();
  }

  surface(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surface();
  }

  surfaceDim(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceDim();
  }

  surfaceBright(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceBright();
  }

  surfaceContainerLowest(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceContainerLowest();
  }

  surfaceContainerLow(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceContainerLow();
  }

  surfaceContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceContainer();
  }

  surfaceContainerHigh(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceContainerHigh();
  }

  surfaceContainerHighest(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceContainerHighest();
  }

  onSurface(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSurface();
  }

  surfaceVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceVariant();
  }

  onSurfaceVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSurfaceVariant();
  }

  outline(): DynamicColor {
    return MaterialDynamicColors.colorSpec.outline();
  }

  outlineVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.outlineVariant();
  }

  inverseSurface(): DynamicColor {
    return MaterialDynamicColors.colorSpec.inverseSurface();
  }

  inverseOnSurface(): DynamicColor {
    return MaterialDynamicColors.colorSpec.inverseOnSurface();
  }

  shadow(): DynamicColor {
    return MaterialDynamicColors.colorSpec.shadow();
  }

  scrim(): DynamicColor {
    return MaterialDynamicColors.colorSpec.scrim();
  }

  surfaceTint(): DynamicColor {
    return MaterialDynamicColors.colorSpec.surfaceTint();
  }

  ////////////////////////////////////////////////////////////////
  // Primaries [P]                                              //
  ////////////////////////////////////////////////////////////////

  primary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.primary();
  }

  primaryDim(): DynamicColor|undefined {
    return MaterialDynamicColors.colorSpec.primaryDim();
  }

  onPrimary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onPrimary();
  }

  primaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.primaryContainer();
  }

  onPrimaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onPrimaryContainer();
  }

  inversePrimary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.inversePrimary();
  }

  /////////////////////////////////////////////////////////////////
  // Primary Fixed [PF]                                          //
  /////////////////////////////////////////////////////////////////

  primaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.primaryFixed();
  }

  primaryFixedDim(): DynamicColor {
    return MaterialDynamicColors.colorSpec.primaryFixedDim();
  }

  onPrimaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onPrimaryFixed();
  }

  onPrimaryFixedVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onPrimaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Secondaries [Q]                                            //
  ////////////////////////////////////////////////////////////////

  secondary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.secondary();
  }

  secondaryDim(): DynamicColor|undefined {
    return MaterialDynamicColors.colorSpec.secondaryDim();
  }

  onSecondary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSecondary();
  }

  secondaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.secondaryContainer();
  }

  onSecondaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSecondaryContainer();
  }

  /////////////////////////////////////////////////////////////////
  // Secondary Fixed [QF]                                        //
  /////////////////////////////////////////////////////////////////

  secondaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.secondaryFixed();
  }

  secondaryFixedDim(): DynamicColor {
    return MaterialDynamicColors.colorSpec.secondaryFixedDim();
  }

  onSecondaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSecondaryFixed();
  }

  onSecondaryFixedVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onSecondaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Tertiaries [T]                                             //
  ////////////////////////////////////////////////////////////////

  tertiary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.tertiary();
  }

  tertiaryDim(): DynamicColor|undefined {
    return MaterialDynamicColors.colorSpec.tertiaryDim();
  }

  onTertiary(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onTertiary();
  }

  tertiaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.tertiaryContainer();
  }

  onTertiaryContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onTertiaryContainer();
  }

  /////////////////////////////////////////////////////////////////
  // Tertiary Fixed [TF]                                         //
  /////////////////////////////////////////////////////////////////

  tertiaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.tertiaryFixed();
  }

  tertiaryFixedDim(): DynamicColor {
    return MaterialDynamicColors.colorSpec.tertiaryFixedDim();
  }

  onTertiaryFixed(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onTertiaryFixed();
  }

  onTertiaryFixedVariant(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onTertiaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Errors [E]                                                 //
  ////////////////////////////////////////////////////////////////

  error(): DynamicColor {
    return MaterialDynamicColors.colorSpec.error();
  }

  errorDim(): DynamicColor|undefined {
    return MaterialDynamicColors.colorSpec.errorDim();
  }

  onError(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onError();
  }

  errorContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.errorContainer();
  }

  onErrorContainer(): DynamicColor {
    return MaterialDynamicColors.colorSpec.onErrorContainer();
  }

  ////////////////////////////////////////////////////////////////
  // All Colors                                                 //
  ////////////////////////////////////////////////////////////////

  allColors: DynamicColor[] = [
    this.background(),
    this.onBackground(),
    this.surface(),
    this.surfaceDim(),
    this.surfaceBright(),
    this.surfaceContainerLowest(),
    this.surfaceContainerLow(),
    this.surfaceContainer(),
    this.surfaceContainerHigh(),
    this.surfaceContainerHighest(),
    this.onSurface(),
    this.onSurfaceVariant(),
    this.outline(),
    this.outlineVariant(),
    this.inverseSurface(),
    this.inverseOnSurface(),
    this.primary(),
    this.primaryDim(),
    this.onPrimary(),
    this.primaryContainer(),
    this.onPrimaryContainer(),
    this.primaryFixed(),
    this.primaryFixedDim(),
    this.onPrimaryFixed(),
    this.onPrimaryFixedVariant(),
    this.inversePrimary(),
    this.secondary(),
    this.secondaryDim(),
    this.onSecondary(),
    this.secondaryContainer(),
    this.onSecondaryContainer(),
    this.secondaryFixed(),
    this.secondaryFixedDim(),
    this.onSecondaryFixed(),
    this.onSecondaryFixedVariant(),
    this.tertiary(),
    this.tertiaryDim(),
    this.onTertiary(),
    this.tertiaryContainer(),
    this.onTertiaryContainer(),
    this.tertiaryFixed(),
    this.tertiaryFixedDim(),
    this.onTertiaryFixed(),
    this.onTertiaryFixedVariant(),
    this.error(),
    this.errorDim(),
    this.onError(),
    this.errorContainer(),
    this.onErrorContainer(),
  ].filter((c) => c !== undefined);

  // Static variables are deprecated. Use the instance methods to get correct
  // specs based on request.

  /** @deprecated Use highestSurface() instead. */
  static highestSurface(s: DynamicScheme): DynamicColor {
    return MaterialDynamicColors.colorSpec.highestSurface(s);
  }

  /** @deprecated Use primaryPaletteKeyColor() instead. */
  static primaryPaletteKeyColor =
      MaterialDynamicColors.colorSpec.primaryPaletteKeyColor();

  /** @deprecated Use secondaryPaletteKeyColor() instead. */
  static secondaryPaletteKeyColor =
      MaterialDynamicColors.colorSpec.secondaryPaletteKeyColor();

  /** @deprecated Use tertiaryPaletteKeyColor() instead. */
  static tertiaryPaletteKeyColor =
      MaterialDynamicColors.colorSpec.tertiaryPaletteKeyColor();

  /** @deprecated Use neutralPaletteKeyColor() instead. */
  static neutralPaletteKeyColor =
      MaterialDynamicColors.colorSpec.neutralPaletteKeyColor();

  /** @deprecated Use neutralVariantPaletteKeyColor() instead. */
  static neutralVariantPaletteKeyColor =
      MaterialDynamicColors.colorSpec.neutralVariantPaletteKeyColor();

  /** @deprecated Use background() instead. */
  static background = MaterialDynamicColors.colorSpec.background();

  /** @deprecated Use background() instead. */
  static onBackground = MaterialDynamicColors.colorSpec.onBackground();

  /** @deprecated Use surface() instead. */
  static surface = MaterialDynamicColors.colorSpec.surface();

  /** @deprecated Use surfaceDim() instead. */
  static surfaceDim = MaterialDynamicColors.colorSpec.surfaceDim();

  /** @deprecated Use surfaceBright() instead. */
  static surfaceBright = MaterialDynamicColors.colorSpec.surfaceBright();

  /** @deprecated Use surfaceContainerLowest() instead. */
  static surfaceContainerLowest =
      MaterialDynamicColors.colorSpec.surfaceContainerLowest();

  /** @deprecated Use surfaceContainerLow() instead. */
  static surfaceContainerLow =
      MaterialDynamicColors.colorSpec.surfaceContainerLow();

  /** @deprecated Use surfaceContainer() instead. */
  static surfaceContainer = MaterialDynamicColors.colorSpec.surfaceContainer();

  /** @deprecated Use surfaceContainerHigh() instead. */
  static surfaceContainerHigh =
      MaterialDynamicColors.colorSpec.surfaceContainerHigh();

  /** @deprecated Use surfaceContainerHighest() instead. */
  static surfaceContainerHighest =
      MaterialDynamicColors.colorSpec.surfaceContainerHighest();

  /** @deprecated Use onSurface() instead. */
  static onSurface = MaterialDynamicColors.colorSpec.onSurface();

  /** @deprecated Use surfaceVariant() instead. */
  static surfaceVariant = MaterialDynamicColors.colorSpec.surfaceVariant();

  /** @deprecated Use onSurfaceVariant() instead. */
  static onSurfaceVariant = MaterialDynamicColors.colorSpec.onSurfaceVariant();

  /** @deprecated Use inverseSurface() instead. */
  static inverseSurface = MaterialDynamicColors.colorSpec.inverseSurface();

  /** @deprecated Use inverseOnSurface() instead. */
  static inverseOnSurface = MaterialDynamicColors.colorSpec.inverseOnSurface();

  /** @deprecated Use outline() instead. */
  static outline = MaterialDynamicColors.colorSpec.outline();

  /** @deprecated Use outlineVariant() instead. */
  static outlineVariant = MaterialDynamicColors.colorSpec.outlineVariant();

  /** @deprecated Use shadow() instead. */
  static shadow = MaterialDynamicColors.colorSpec.shadow();

  /** @deprecated Use scrim() instead. */
  static scrim = MaterialDynamicColors.colorSpec.scrim();

  /** @deprecated Use surfaceTint() instead. */
  static surfaceTint = MaterialDynamicColors.colorSpec.surfaceTint();

  /** @deprecated Use primary() instead. */
  static primary = MaterialDynamicColors.colorSpec.primary();

  /** @deprecated Use onPrimary() instead. */
  static onPrimary = MaterialDynamicColors.colorSpec.onPrimary();

  /** @deprecated Use primaryContainer() instead. */
  static primaryContainer = MaterialDynamicColors.colorSpec.primaryContainer();

  /** @deprecated Use onPrimaryContainer() instead. */
  static onPrimaryContainer =
      MaterialDynamicColors.colorSpec.onPrimaryContainer();

  /** @deprecated Use inversePrimary() instead. */
  static inversePrimary = MaterialDynamicColors.colorSpec.inversePrimary();

  /** @deprecated Use secondary() instead. */
  static secondary = MaterialDynamicColors.colorSpec.secondary();

  /** @deprecated Use onSecondary() instead. */
  static onSecondary = MaterialDynamicColors.colorSpec.onSecondary();

  /** @deprecated Use secondaryContainer() instead. */
  static secondaryContainer =
      MaterialDynamicColors.colorSpec.secondaryContainer();

  /** @deprecated Use onSecondaryContainer() instead. */
  static onSecondaryContainer =
      MaterialDynamicColors.colorSpec.onSecondaryContainer();

  /** @deprecated Use tertiary() instead. */
  static tertiary = MaterialDynamicColors.colorSpec.tertiary();

  /** @deprecated Use onTertiary() instead. */
  static onTertiary = MaterialDynamicColors.colorSpec.onTertiary();

  /** @deprecated Use tertiaryContainer() instead. */
  static tertiaryContainer = MaterialDynamicColors.colorSpec.tertiaryContainer();

  /** @deprecated Use onTertiaryContainer() instead. */
  static onTertiaryContainer =
      MaterialDynamicColors.colorSpec.onTertiaryContainer();

  /** @deprecated Use error() instead. */
  static error = MaterialDynamicColors.colorSpec.error();

  /** @deprecated Use onError() instead. */
  static onError = MaterialDynamicColors.colorSpec.onError();

  /** @deprecated Use errorContainer() instead. */
  static errorContainer = MaterialDynamicColors.colorSpec.errorContainer();

  /** @deprecated Use onErrorContainer() instead. */
  static onErrorContainer = MaterialDynamicColors.colorSpec.onErrorContainer();

  /** @deprecated Use primaryFixed() instead. */
  static primaryFixed = MaterialDynamicColors.colorSpec.primaryFixed();

  /** @deprecated Use primaryFixedDim() instead. */
  static primaryFixedDim = MaterialDynamicColors.colorSpec.primaryFixedDim();

  /** @deprecated Use onPrimaryFixed() instead. */
  static onPrimaryFixed = MaterialDynamicColors.colorSpec.onPrimaryFixed();

  /** @deprecated Use onPrimaryFixedVariant() instead. */
  static onPrimaryFixedVariant =
      MaterialDynamicColors.colorSpec.onPrimaryFixedVariant();

  /** @deprecated Use secondaryFixed() instead. */
  static secondaryFixed = MaterialDynamicColors.colorSpec.secondaryFixed();

  /** @deprecated Use secondaryFixedDim() instead. */
  static secondaryFixedDim = MaterialDynamicColors.colorSpec.secondaryFixedDim();

  /** @deprecated Use onSecondaryFixed() instead. */
  static onSecondaryFixed = MaterialDynamicColors.colorSpec.onSecondaryFixed();

  /** @deprecated Use onSecondaryFixedVariant() instead. */
  static onSecondaryFixedVariant =
      MaterialDynamicColors.colorSpec.onSecondaryFixedVariant();

  /** @deprecated Use tertiaryFixed() instead. */
  static tertiaryFixed = MaterialDynamicColors.colorSpec.tertiaryFixed();

  /** @deprecated Use tertiaryFixedDim() instead. */
  static tertiaryFixedDim = MaterialDynamicColors.colorSpec.tertiaryFixedDim();

  /** @deprecated Use onTertiaryFixed() instead. */
  static onTertiaryFixed = MaterialDynamicColors.colorSpec.onTertiaryFixed();

  /** @deprecated Use onTertiaryFixedVariant() instead. */
  static onTertiaryFixedVariant =
      MaterialDynamicColors.colorSpec.onTertiaryFixedVariant();
}

File: java/dynamiccolor/MaterialDynamicColors.java
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dynamiccolor;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.Arrays;
import java.util.List;
import java.util.function.Supplier;

/** Named colors, otherwise known as tokens, or roles, in the Material Design system. */
public final class MaterialDynamicColors {

  private static final ColorSpec colorSpec = new ColorSpec2026();

  @NonNull
  public DynamicColor highestSurface(@NonNull DynamicScheme s) {
    return colorSpec.highestSurface(s);
  }

  ////////////////////////////////////////////////////////////////
  // Main Palettes                                              //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor primaryPaletteKeyColor() {
    return colorSpec.primaryPaletteKeyColor();
  }

  @NonNull
  public DynamicColor secondaryPaletteKeyColor() {
    return colorSpec.secondaryPaletteKeyColor();
  }

  @NonNull
  public DynamicColor tertiaryPaletteKeyColor() {
    return colorSpec.tertiaryPaletteKeyColor();
  }

  @NonNull
  public DynamicColor neutralPaletteKeyColor() {
    return colorSpec.neutralPaletteKeyColor();
  }

  @NonNull
  public DynamicColor neutralVariantPaletteKeyColor() {
    return colorSpec.neutralVariantPaletteKeyColor();
  }

  @NonNull
  public DynamicColor errorPaletteKeyColor() {
    return colorSpec.errorPaletteKeyColor();
  }

  ////////////////////////////////////////////////////////////////
  // Surfaces [S]                                               //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor background() {
    return colorSpec.background();
  }

  @NonNull
  public DynamicColor onBackground() {
    return colorSpec.onBackground();
  }

  @NonNull
  public DynamicColor surface() {
    return colorSpec.surface();
  }

  @NonNull
  public DynamicColor surfaceDim() {
    return colorSpec.surfaceDim();
  }

  @NonNull
  public DynamicColor surfaceBright() {
    return colorSpec.surfaceBright();
  }

  @NonNull
  public DynamicColor surfaceContainerLowest() {
    return colorSpec.surfaceContainerLowest();
  }

  @NonNull
  public DynamicColor surfaceContainerLow() {
    return colorSpec.surfaceContainerLow();
  }

  @NonNull
  public DynamicColor surfaceContainer() {
    return colorSpec.surfaceContainer();
  }

  @NonNull
  public DynamicColor surfaceContainerHigh() {
    return colorSpec.surfaceContainerHigh();
  }

  @NonNull
  public DynamicColor surfaceContainerHighest() {
    return colorSpec.surfaceContainerHighest();
  }

  @NonNull
  public DynamicColor onSurface() {
    return colorSpec.onSurface();
  }

  @NonNull
  public DynamicColor surfaceVariant() {
    return colorSpec.surfaceVariant();
  }

  @NonNull
  public DynamicColor onSurfaceVariant() {
    return colorSpec.onSurfaceVariant();
  }

  @NonNull
  public DynamicColor inverseSurface() {
    return colorSpec.inverseSurface();
  }

  @NonNull
  public DynamicColor inverseOnSurface() {
    return colorSpec.inverseOnSurface();
  }

  @NonNull
  public DynamicColor outline() {
    return colorSpec.outline();
  }

  @NonNull
  public DynamicColor outlineVariant() {
    return colorSpec.outlineVariant();
  }

  @NonNull
  public DynamicColor shadow() {
    return colorSpec.shadow();
  }

  @NonNull
  public DynamicColor scrim() {
    return colorSpec.scrim();
  }

  @NonNull
  public DynamicColor surfaceTint() {
    return colorSpec.surfaceTint();
  }

  ////////////////////////////////////////////////////////////////
  // Primaries [P]                                              //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor primary() {
    return colorSpec.primary();
  }

  @Nullable
  public DynamicColor primaryDim() {
    return colorSpec.primaryDim();
  }

  @NonNull
  public DynamicColor onPrimary() {
    return colorSpec.onPrimary();
  }

  @NonNull
  public DynamicColor primaryContainer() {
    return colorSpec.primaryContainer();
  }

  @NonNull
  public DynamicColor onPrimaryContainer() {
    return colorSpec.onPrimaryContainer();
  }

  @NonNull
  public DynamicColor inversePrimary() {
    return colorSpec.inversePrimary();
  }

  /////////////////////////////////////////////////////////////////
  // Primary Fixed Colors [PF]                                   //
  /////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor primaryFixed() {
    return colorSpec.primaryFixed();
  }

  @NonNull
  public DynamicColor primaryFixedDim() {
    return colorSpec.primaryFixedDim();
  }

  @NonNull
  public DynamicColor onPrimaryFixed() {
    return colorSpec.onPrimaryFixed();
  }

  @NonNull
  public DynamicColor onPrimaryFixedVariant() {
    return colorSpec.onPrimaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Secondaries [Q]                                            //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor secondary() {
    return colorSpec.secondary();
  }

  @Nullable
  public DynamicColor secondaryDim() {
    return colorSpec.secondaryDim();
  }

  @NonNull
  public DynamicColor onSecondary() {
    return colorSpec.onSecondary();
  }

  @NonNull
  public DynamicColor secondaryContainer() {
    return colorSpec.secondaryContainer();
  }

  @NonNull
  public DynamicColor onSecondaryContainer() {
    return colorSpec.onSecondaryContainer();
  }

  /////////////////////////////////////////////////////////////////
  // Secondary Fixed Colors [QF]                                 //
  /////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor secondaryFixed() {
    return colorSpec.secondaryFixed();
  }

  @NonNull
  public DynamicColor secondaryFixedDim() {
    return colorSpec.secondaryFixedDim();
  }

  @NonNull
  public DynamicColor onSecondaryFixed() {
    return colorSpec.onSecondaryFixed();
  }

  @NonNull
  public DynamicColor onSecondaryFixedVariant() {
    return colorSpec.onSecondaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Tertiaries [T]                                             //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor tertiary() {
    return colorSpec.tertiary();
  }

  @Nullable
  public DynamicColor tertiaryDim() {
    return colorSpec.tertiaryDim();
  }

  @NonNull
  public DynamicColor onTertiary() {
    return colorSpec.onTertiary();
  }

  @NonNull
  public DynamicColor tertiaryContainer() {
    return colorSpec.tertiaryContainer();
  }

  @NonNull
  public DynamicColor onTertiaryContainer() {
    return colorSpec.onTertiaryContainer();
  }

  /////////////////////////////////////////////////////////////////
  // Tertiary Fixed Colors [TF]                                  //
  /////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor tertiaryFixed() {
    return colorSpec.tertiaryFixed();
  }

  @NonNull
  public DynamicColor tertiaryFixedDim() {
    return colorSpec.tertiaryFixedDim();
  }

  @NonNull
  public DynamicColor onTertiaryFixed() {
    return colorSpec.onTertiaryFixed();
  }

  @NonNull
  public DynamicColor onTertiaryFixedVariant() {
    return colorSpec.onTertiaryFixedVariant();
  }

  ////////////////////////////////////////////////////////////////
  // Errors [E]                                                 //
  ////////////////////////////////////////////////////////////////

  @NonNull
  public DynamicColor error() {
    return colorSpec.error();
  }

  @Nullable
  public DynamicColor errorDim() {
    return colorSpec.errorDim();
  }

  @NonNull
  public DynamicColor onError() {
    return colorSpec.onError();
  }

  @NonNull
  public DynamicColor errorContainer() {
    return colorSpec.errorContainer();
  }

  @NonNull
  public DynamicColor onErrorContainer() {
    return colorSpec.onErrorContainer();
  }

  ////////////////////////////////////////////////////////////////
  // Android-only colors                                        //
  ////////////////////////////////////////////////////////////////

  /**
   * These colors were present in Android framework before Android U, and used by MDC controls. They
   * should be avoided, if possible. It's unclear if they're used on multiple backgrounds, and if
   * they are, they can't be adjusted for contrast.* For now, they will be set with no background,
   * and those won't adjust for contrast, avoiding issues.
   *
   * <p>* For example, if the same color is on a white background _and_ black background, there's no
   * way to increase contrast with either without losing contrast with the other.
   */

  // colorControlActivated documented as colorAccent in M3 & GM3.
  // colorAccent documented as colorSecondary in M3 and colorPrimary in GM3.
  // Android used Material's Container as Primary/Secondary/Tertiary at launch.
  // Therefore, this is a duplicated version of Primary Container.
  @NonNull
  public DynamicColor controlActivated() {
    return colorSpec.controlActivated();
  }

  // colorControlNormal documented as textColorSecondary in M3 & GM3.
  // In Material, textColorSecondary points to onSurfaceVariant in the non-disabled state,
  // which is Neutral Variant T30/80 in light/dark.
  @NonNull
  public DynamicColor controlNormal() {
    return colorSpec.controlNormal();
  }

  // colorControlHighlight documented, in both M3 & GM3:
  // Light mode: #1f000000 dark mode: #33ffffff.
  // These are black and white with some alpha.
  // 1F hex = 31 decimal; 31 / 255 = 12% alpha.
  // 33 hex = 51 decimal; 51 / 255 = 20% alpha.
  // DynamicColors do not support alpha currently, and _may_ not need it for this use case,
  // depending on how MDC resolved alpha for the other cases.
  // Returning black in dark mode, white in light mode.
  @NonNull
  public DynamicColor controlHighlight() {
    return colorSpec.controlHighlight();
  }

  // textColorPrimaryInverse documented, in both M3 & GM3, documented as N10/N90.
  @NonNull
  public DynamicColor textPrimaryInverse() {
    return colorSpec.textPrimaryInverse();
  }

  // textColorSecondaryInverse and textColorTertiaryInverse both documented, in both M3 & GM3, as
  // NV30/NV80
  @NonNull
  public DynamicColor textSecondaryAndTertiaryInverse() {
    return colorSpec.textSecondaryAndTertiaryInverse();
  }

  // textColorPrimaryInverseDisableOnly documented, in both M3 & GM3, as N10/N90
  @NonNull
  public DynamicColor textPrimaryInverseDisableOnly() {
    return colorSpec.textPrimaryInverseDisableOnly();
  }

  // textColorSecondaryInverse and textColorTertiaryInverse in disabled state both documented,
  // in both M3 & GM3, as N10/N90
  @NonNull
  public DynamicColor textSecondaryAndTertiaryInverseDisabled() {
    return colorSpec.textSecondaryAndTertiaryInverseDisabled();
  }

  // textColorHintInverse documented, in both M3 & GM3, as N10/N90
  @NonNull
  public DynamicColor textHintInverse() {
    return colorSpec.textHintInverse();
  }

  ////////////////////////////////////////////////////////////////
  // All Colors                                                 //
  ////////////////////////////////////////////////////////////////

  /** All dynamic colors in Material Design system. */
  public final List<Supplier<DynamicColor>> allDynamicColors() {
    return Arrays.asList(
        this::primaryPaletteKeyColor,
        this::secondaryPaletteKeyColor,
        this::tertiaryPaletteKeyColor,
        this::neutralPaletteKeyColor,
        this::neutralVariantPaletteKeyColor,
        this::errorPaletteKeyColor,
        this::background,
        this::onBackground,
        this::surface,
        this::surfaceDim,
        this::surfaceBright,
        this::surfaceContainerLowest,
        this::surfaceContainerLow,
        this::surfaceContainer,
        this::surfaceContainerHigh,
        this::surfaceContainerHighest,
        this::onSurface,
        this::surfaceVariant,
        this::onSurfaceVariant,
        this::outline,
        this::outlineVariant,
        this::inverseSurface,
        this::inverseOnSurface,
        this::shadow,
        this::scrim,
        this::surfaceTint,
        this::primary,
        this::primaryDim,
        this::onPrimary,
        this::primaryContainer,
        this::onPrimaryContainer,
        this::primaryFixed,
        this::primaryFixedDim,
        this::onPrimaryFixed,
        this::onPrimaryFixedVariant,
        this::inversePrimary,
        this::secondary,
        this::secondaryDim,
        this::onSecondary,
        this::secondaryContainer,
        this::onSecondaryContainer,
        this::secondaryFixed,
        this::secondaryFixedDim,
        this::onSecondaryFixed,
        this::onSecondaryFixedVariant,
        this::tertiary,
        this::tertiaryDim,
        this::onTertiary,
        this::tertiaryContainer,
        this::onTertiaryContainer,
        this::tertiaryFixed,
        this::tertiaryFixedDim,
        this::onTertiaryFixed,
        this::onTertiaryFixedVariant,
        this::error,
        this::errorDim,
        this::onError,
        this::errorContainer,
        this::onErrorContainer,
        this::controlActivated,
        this::controlNormal,
        this::controlHighlight,
        this::textPrimaryInverse,
        this::textSecondaryAndTertiaryInverse,
        this::textPrimaryInverseDisableOnly,
        this::textSecondaryAndTertiaryInverseDisabled,
        this::textHintInverse);
  }
}
